crates/wasm-mutate/src/mutators/peephole/eggsy/expr_enumerator.rs

```diff
@@ -0,0 +1,877 @@
+
+use super::lang::Lang;
+
+macro_rules! binop {
+    ($lang:ident, $left:ident, $right: ident, $egraph: ident, $rnd: ident, $depth: ident, $recexpr: ident) => {{
+        let lcope = $left;
+        let rcopy = $right;
+        // Cartesian product of the operands
+        let t = lazy_expand(lcope, $egraph.clone(), $depth, $rnd, &$recexpr)
+            .flat_map(move |e| {
+                std::iter::repeat(e).zip(lazy_expand(
+                    rcopy,
+                    $egraph.clone(),
+                    $depth,
+                    $rnd,
+                    &$recexpr,
+                ))
+            })
+            .map(move |(l, r)| $recexpr.borrow_mut().add(Lang::$lang([l, r])));
+
+        Box::new(t)
+    }};
+}
+
+macro_rules! unop {
+    ($lang:ident,$arg: ident, $egraph: ident, $rnd: ident, $depth: ident, $recexpr: ident) => {{
+        let lcope = $arg;
+        let t = lazy_expand(lcope, $egraph, $depth, $rnd, &$recexpr)
+            .map(move |l| $recexpr.borrow_mut().add(Lang::$lang([l])));
+
+        Box::new(t)
+    }};
+}
+
+macro_rules! store {
+    ($lang:ident, $left:ident, $right: ident, $egraph: ident, $rnd: ident, $depth: ident, $recexpr: ident, $align: ident, $offset: ident, $mem: ident) => {{
+        let lcope = $left;
+        let rcopy = $right;
+        let t = lazy_expand(lcope, $egraph.clone(), $depth, $rnd, &$recexpr)
+            .flat_map(move |e| {
+                std::iter::repeat(e).zip(lazy_expand(
+                    rcopy,
+                    $egraph.clone(),
+                    $depth,
+                    $rnd,
+                    &$recexpr,
+                ))
+            })
+            .map(move |(l, r)| {
+                $recexpr.borrow_mut().add(Lang::$lang {
+                    value_and_offset: [l, r],
+                    align: $align,
+                    mem: $mem,
+                    static_offset: $offset,
+                })
+            });
+
+        Box::new(t)
+    }};
+}
+
+macro_rules! load {
+    ($lang:ident, $arg: ident, $egraph: ident, $rnd: ident, $depth: ident, $recexpr: ident, $align: ident, $offset: ident, $mem: ident) => {{
+        let lcope = $arg;
+        let t = lazy_expand(lcope, $egraph, $depth, $rnd, &$recexpr).map(move |l| {
+            $recexpr.borrow_mut().add(Lang::$lang {
+                offset: l,
+                static_offset: $offset,
+                align: $align,
+                mem: $mem,
+            })
+        });
+
+        Box::new(t)
+    }};
+}
+
+macro_rules! local_or_global {
+    ($lang:ident, $idx: ident, $arg: ident, $egraph: ident, $rnd: ident, $depth: ident, $recexpr: ident) => {{
+        let lcope = $arg;
+        let t = lazy_expand(lcope, $egraph, $depth, $rnd, &$recexpr)
+            .map(move |l| $recexpr.borrow_mut().add(Lang::$lang($idx, l)));
+
+        Box::new(t)
+    }};
+}
+
+/// Returns a lazy iterator over random expressions in the eggraph
+///
+pub fn lazy_expand<'a>(
+    id: Id,
+    egraph: EG,
```

Comment:
Can we make this parameter `&'a EG` so that we don't have to clone the whole e-graph below?

That does mean that the returned iterator borrows the e-graph, but it is already `+ 'a` so that shouldn't be a problem.